app/controllers/active_storage/blobs/proxy_controller.rb

```diff
@@ -22,6 +26,38 @@ def show
 private
 
   def authorize_blob
-    redirect_to "/sign-in" unless user_signed_in?
+    return redirect_to "/sign-in" unless user_signed_in?
+
+    if related_investigation
+      if attachment_is_correspondence_or_non_image_investigation_attachment? && !InvestigationPolicy.new(current_user, related_investigation).view_protected_details?
+        return redirect_to "/", flash: { warning: "Not authorized to view this attachment" }
+      end
+
+      return redirect_to "/", flash: { warning: "Not authorized to view this attachment" } unless InvestigationPolicy.new(current_user, related_investigation).view_non_protected_details?
+    end
+  end
+
+  def attachment_categorizer
+    AttachmentCategorizer.new(@blob)
+  end
+
+  def related_investigation
+    attachment_categorizer.related_investigation
+  end
+
+  def non_activity_attachment
+    attachment_categorizer.non_activity_attachment
+  end
+
+  def is_an_image?
+    attachment_categorizer.is_an_image?
+  end
+
+  def is_an_investigation_image?
+    non_activity_attachment.record_type == "Investigation" && is_an_image?
+  end
+
+  def attachment_is_correspondence_or_non_image_investigation_attachment?
```

Comment:
Suggestion: name the method with a higher level meaning. The implementation is clearer.

```suggestion
  def attachment_is_protected_type?
```